README.rst
freesurfer-stats
================

.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
   :target: https://github.com/psf/black
.. image:: https://github.com/fphammerle/freesurfer-stats/workflows/tests/badge.svg
   :target: https://github.com/fphammerle/freesurfer-stats/actions
.. image:: https://coveralls.io/repos/github/fphammerle/freesurfer-stats/badge.svg?branch=master
   :target: https://coveralls.io/github/fphammerle/freesurfer-stats?branch=master
.. image:: https://img.shields.io/pypi/v/freesurfer-stats.svg
   :target: https://pypi.org/project/freesurfer-stats/#history
.. image:: https://img.shields.io/pypi/pyversions/freesurfer-stats.svg
   :target: https://pypi.org/project/freesurfer-stats/
.. image:: https://zenodo.org/badge/194054168.svg
   :target: https://zenodo.org/badge/latestdoi/194054168

Python Library to Read FreeSurfer's Cortical Parcellation Anatomical Statistics
`subject/stats/[rl]h.aparc.*.stats`

Freesurfer https://surfer.nmr.mgh.harvard.edu/

Install
-------

.. code:: sh

    pip3 install --user freesurfer-stats

Releases follow the `semantic versioning <https://semver.org/>` scheme.

Usage
-----

.. code:: python

    >>> from freesurfer_stats import CorticalParcellationStats
    >>> stats = CorticalParcellationStats.read('tests/subjects/fabian/stats/lh.aparc.DKTatlas.stats')
    >>> stats.headers['subjectname']
    'fabian'
    >>> stats.headers['CreationTime'].isoformat()
    '2019-05-09T21:05:54+00:00'
    >>> stats.headers['cvs_version']
    'Id: mris_anatomical_stats.c,v 1.79 2016/03/14 15:15:34 greve Exp'
    >>> stats.headers['cmdline'][:64]
    'mris_anatomical_stats -th3 -mgz -cortex ../label/lh.cortex.label'
    >>> stats.hemisphere
    'left'
    >>> stats.whole_brain_measurements['estimated_total_intracranial_volume_mm^3']
    0    1.670487e+06
    Name: estimated_total_intracranial_volume_mm^3, dtype: float64
    >>> stats.whole_brain_measurements['white_surface_total_area_mm^2']
    0    98553
    Name: white_surface_total_area_mm^2, dtype: int64
    >>> stats.structural_measurements[['structure_name', 'surface_area_mm^2',
    ...                                'gray_matter_volume_mm^3']].head()
                structure_name  surface_area_mm^2  gray_matter_volume_mm^3
    0  caudalanteriorcingulate               1472                     4258
    1      caudalmiddlefrontal               3039                     8239
    2                   cuneus               2597                     6722
    3               entorhinal                499                     2379
    4                 fusiform               3079                     9064

Load Multiple Stats Files
~~~~~~~~~~~~~~~~~~~~~~~~~

.. code:: python

    >>> import glob, pandas
    >>> from freesurfer_stats import CorticalParcellationStats
    >>> def load_whole_brain_measurements(stats_path) -> pandas.DataFrame:
    ...     stats = CorticalParcellationStats.read(stats_path)
    ...     stats.whole_brain_measurements['subject'] = stats.headers['subjectname']
    ...     stats.whole_brain_measurements['source_basename'] = os.path.basename(stats_path)
    ...     stats.whole_brain_measurements['hemisphere'] = stats.hemisphere
    ...     return stats.whole_brain_measurements
    ...
    >>> whole_brain_measurements = pandas.concat(
    ...     map(load_whole_brain_measurements, glob.glob('tests/subjects/fabian/stats/*h.aparc*.stats')),
    ...     sort=False)
    >>> whole_brain_measurements.reset_index(drop=True, inplace=True)
    >>> whole_brain_measurements[['subject', 'source_basename', 'hemisphere',
    ...                           'white_surface_total_area_mm^2', 'pial_surface_total_area_mm^2']]
      subject          source_basename hemisphere  white_surface_total_area_mm^2  pial_surface_total_area_mm^2
    0  fabian  lh.aparc.DKTatlas.stats       left                        98553.0                           NaN
    1  fabian           rh.aparc.stats      right                        99468.9                           NaN
    2  fabian    rh.aparc.a2009s.stats      right                        99494.9                           NaN
    3  fabian  rh.aparc.DKTatlas.stats      right                        99494.9                           NaN
    4  fabian           lh.aparc.stats       left                        98536.5                           NaN
    5  fabian      lh.aparc.pial.stats       left                            NaN                      118601.0
    6  fabian      rh.aparc.pial.stats      right                            NaN                      121260.0
    7  fabian    lh.aparc.a2009s.stats       left                        98553.0                           NaN

Load Stats File From Webserver, Amazon S3 or Google Cloud Storage
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code:: python

    >>> from freesurfer_stats import CorticalParcellationStats
    >>> stats = CorticalParcellationStats.read('https://[...]/stats/rh.aparc.stats')
    >>> stats.whole_brain_measurements['total_cortical_gray_matter_volume_mm^3']
    0    553998.311189
    Name: total_cortical_gray_matter_volume_mm^3, dtype: float64

Replace ``https://`` with ``s3://`` or ``gcs://``.

Credentials for S3 may be provided in ``~/.aws/credentials``
or via environment variables.
See `S3Fs docs <https://s3fs.readthedocs.io/en/latest/#credentials>`__.

Tests
-----

.. code:: sh

    pip3 install --user pipenv
    git clone https://github.com/fphammerle/freesurfer-stats.git
    cd freesurfer-stats
    pipenv sync --dev
    pipenv run pylint freesurfer_stats
    pipenv run pytest


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Removed
- compatibility with `python3.5` & `python3.6`
- compatibility with `pandas<0.23`

## [1.2.1] - 2020-12-31
### Fixed
- `AttributeError` when using pandas v1.2.0

## [1.2.0] - 2020-05-30
### Added
- `CorticalParcellationStats.read` support `pathlib.Path`, `"http://…"`, `"https://…"`, `"s3://…"` etc.
  via `pandas.io.common.get_filepath_or_buffer`
  (https://github.com/fphammerle/freesurfer-stats/issues/6)

## [1.1.1] - 2020-05-07
### Fixed
- fixed parsing of `BrainVolStatsFixed` header
  (https://github.com/fphammerle/freesurfer-stats/pull/1 @soichih,
  https://github.com/fphammerle/freesurfer-stats/pull/9)

## [1.1.0] - 2020-05-06
### Added
- compatibility with pandas v1

## [1.0.0] - 2019-06-27

[Unreleased]: https://github.com/fphammerle/freesurfer-stats/compare/1.2.1...HEAD
[1.2.1]: https://github.com/fphammerle/freesurfer-stats/compare/1.2.0...1.2.1
[1.2.0]: https://github.com/fphammerle/freesurfer-stats/compare/1.1.1...1.2.0
[1.1.1]: https://github.com/fphammerle/freesurfer-stats/compare/1.1.0...1.1.1
[1.1.0]: https://github.com/fphammerle/freesurfer-stats/compare/1.0.0...1.1.0
[1.0.0]: https://github.com/fphammerle/freesurfer-stats/tree/1.0.0


tests/conftest.py
"""
freesurfer-stats, a Python Library to Read FreeSurfer's Cortical Parcellation Anatomical Statistics
Copyright (C) 2019 Fabian Peter Hammerle <fabian@hammerle.me>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""
import os

SUBJECTS_DIR = os.path.join(os.path.dirname(__file__), "subjects")


tests/test_cortical_parcellation_stats.py
"""
freesurfer-stats, a Python Library to Read FreeSurfer's Cortical Parcellation Anatomical Statistics
Copyright (C) 2019 Fabian Peter Hammerle <fabian@hammerle.me>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""
import datetime
import os
import pathlib

import numpy
import pandas.util.testing
import pytest

from conftest import SUBJECTS_DIR
from freesurfer_stats import CorticalParcellationStats

# pylint: disable=too-many-arguments


@pytest.mark.parametrize(
    (
        "path",
        "headers",
        "hemisphere",
        "whole_brain_measurements",
        "structural_measurements_length",
        "structural_measurements_subset",
    ),
    [
        (
            os.path.join(
                SUBJECTS_DIR, "fabian", "stats", "lh.aparc.DKTatlas.stats.short"
            ),
            {
                "CreationTime": datetime.datetime(
                    2019, 5, 9, 21, 5, 54, tzinfo=datetime.timezone.utc
                ),
                "generating_program": "mris_anatomical_stats",
                "cvs_version": "Id: mris_anatomical_stats.c,v 1.79"
                " 2016/03/14 15:15:34 greve Exp",
                "mrisurf.c-cvs_version": "Id: mrisurf.c,v 1.781.2.6"
                " 2016/12/27 16:47:14 zkaufman Exp",
                "cmdline": "mris_anatomical_stats -th3 -mgz -cortex ../label/lh.cortex.label"
                " -f ../stats/lh.aparc.DKTatlas.stats -b -a ../label/lh.aparc.DKTatlas.annot"
                " -c ../label/aparc.annot.DKTatlas.ctab fabian lh white",
                "sysname": "Linux",
                "hostname": "another-hostname",
                "machine": "x86_64",
                "user": "some-username",
                "SUBJECTS_DIR": "/home/some-username/freesurfer-subjects",
                "anatomy_type": "surface",
                "subjectname": "fabian",
                "hemi": "lh",
                "AnnotationFile": "../label/lh.aparc.DKTatlas.annot",
                "AnnotationFileTimeStamp": datetime.datetime(2019, 5, 9, 23, 5, 40),
            },
            "left",
            {
                "white_surface_total_area_mm^2": 98553,
                "mean_thickness_mm": 2.50462,
                "brain_segmentation_volume_mm^3": 1327432.000000,
                "brain_segmentation_volume_without_ventricles_mm^3": 1316285.000000,
                "brain_segmentation_volume_without_ventricles_from_surf_mm^3": 1315572.548920,
                "total_cortical_gray_matter_volume_mm^3": 553998.311189,
                "supratentorial_volume_mm^3": 1172669.548920,
                "supratentorial_volume_without_ventricles_mm^3": 1164180.548920,
                "estimated_total_intracranial_volume_mm^3": 1670487.274486,
            },
            3,
            [
                {
                    "structure_name": "caudalanteriorcingulate",
                    "number_of_vertices": 2061,
                    "surface_area_mm^2": 1472,
                    "gray_matter_volume_mm^3": 4258,
                    "average_thickness_mm": 2.653,
                    "thickness_stddev_mm": 0.644,
                    "integrated_rectified_mean_curvature_mm^-1": 0.135,
                    "integrated_rectified_gaussian_curvature_mm^-2": 0.020,
                    "folding_index": 27,
                    "intrinsic_curvature_index": 1.6,
                },
                {
                    "structure_name": "caudalmiddlefrontal",
                    "number_of_vertices": 4451,
                    "surface_area_mm^2": 3039,
                    "gray_matter_volume_mm^3": 8239,
                    "average_thickness_mm": 2.456,
                    "thickness_stddev_mm": 0.486,
                    "integrated_rectified_mean_curvature_mm^-1": 0.116,
                    "integrated_rectified_gaussian_curvature_mm^-2": 0.020,
                    "folding_index": 42,
                    "intrinsic_curvature_index": 3.7,
                },
                {
                    "structure_name": "insula",
                    "number_of_vertices": 3439,
                    "surface_area_mm^2": 2304,
                    "gray_matter_volume_mm^3": 7594,
                    "average_thickness_mm": 3.193,
                    "thickness_stddev_mm": 0.620,
                    "integrated_rectified_mean_curvature_mm^-1": 0.116,
                    "integrated_rectified_gaussian_curvature_mm^-2": 0.027,
                    "folding_index": 33,
                    "intrinsic_curvature_index": 3.5,
                },
            ],
        ),
        (
            os.path.join(SUBJECTS_DIR, "fabian", "stats", "rh.aparc.pial.stats.short"),
            {
                "CreationTime": datetime.datetime(
                    2019, 5, 9, 21, 3, 42, tzinfo=datetime.timezone.utc
                ),
                "generating_program": "mris_anatomical_stats",
                "cvs_version": "Id: mris_anatomical_stats.c,v 1.79"
                " 2016/03/14 15:15:34 greve Exp",
                "mrisurf.c-cvs_version": "Id: mrisurf.c,v 1.781.2.6"
                " 2016/12/27 16:47:14 zkaufman Exp",
                "cmdline": "mris_anatomical_stats -th3 -mgz -cortex ../label/rh.cortex.label"
                " -f ../stats/rh.aparc.pial.stats -b -a ../label/rh.aparc.annot"
                " -c ../label/aparc.annot.ctab fabian rh pial",
                "sysname": "Linux",
                "hostname": "some-hostname",
                "machine": "x86_64",
                "user": "some-username",
                "SUBJECTS_DIR": "/home/some-username/freesurfer-subjects",
                "anatomy_type": "surface",
                "subjectname": "fabian",
                "hemi": "rh",
                "AnnotationFile": "../label/rh.aparc.annot",
                "AnnotationFileTimeStamp": datetime.datetime(2019, 5, 9, 22, 27, 28),
            },
            "right",
            {
                "pial_surface_total_area_mm^2": 121260,
                "mean_thickness_mm": 2.4817,
                "brain_segmentation_volume_mm^3": 1327432.000000,
                "brain_segmentation_volume_without_ventricles_mm^3": 1316285.000000,
                "brain_segmentation_volume_without_ventricles_from_surf_mm^3": 1315572.548920,
                "total_cortical_gray_matter_volume_mm^3": 553998.311189,
                "supratentorial_volume_mm^3": 1172669.548920,
                "supratentorial_volume_without_ventricles_mm^3": 1164180.548920,
                "estimated_total_intracranial_volume_mm^3": 1670487.274486,
            },
            2,
            [
                {
                    "structure_name": "bankssts",
                    "number_of_vertices": 1344,
                    "surface_area_mm^2": 825,
                    "gray_matter_volume_mm^3": 2171,
                    "average_thickness_mm": 2.436,
                    "thickness_stddev_mm": 0.381,
                    "integrated_rectified_mean_curvature_mm^-1": 0.115,
                    "integrated_rectified_gaussian_curvature_mm^-2": 0.028,
                    "folding_index": 19,
                    "intrinsic_curvature_index": 1.7,
                },
                {
                    "structure_name": "transversetemporal",
                    "number_of_vertices": 651,
                    "surface_area_mm^2": 545,
                    "gray_matter_volume_mm^3": 1061,
                    "average_thickness_mm": 2.251,
                    "thickness_stddev_mm": 0.317,
                    "integrated_rectified_mean_curvature_mm^-1": 0.110,
                    "integrated_rectified_gaussian_curvature_mm^-2": 0.021,
                    "folding_index": 3,
                    "intrinsic_curvature_index": 0.6,
                },
            ],
        ),
        (
            os.path.join(SUBJECTS_DIR, "soichi", "stats", "rh.aparc.stats"),
            {
                "CreationTime": datetime.datetime(
                    2020, 5, 4, 22, 20, 26, tzinfo=datetime.timezone.utc
                ),
                "AnnotationFileTimeStamp": datetime.datetime(2020, 5, 4, 21, 58, 13),
                "AnnotationFile": "../label/rh.aparc.annot",
                "SUBJECTS_DIR": "/N/dc2/scratch/hayashis/bigred3-workflows"
                "/5eb0689676c10ead933d673c/5eb068b076c10e7b013d673f",
                "anatomy_type": "surface",
                "cmdline": "mris_anatomical_stats -th3 -mgz -cortex ../label/rh.cortex.label "
                "-f ../stats/rh.aparc.stats -b -a ../label/rh.aparc.annot -c "
                "../label/aparc.annot.ctab output rh white",
                "cvs_version": "7.0.0",
                "generating_program": "mris_anatomical_stats",
                "hemi": "rh",
                "hostname": "nid00762",
                "machine": "x86_64",
                "mrisurf.c-cvs_version": "7.0.0",
                "subjectname": "output",
                "sysname": "Linux",
                "user": "hayashis",
                "BrainVolStatsFixed": "NotNeeded because voxelvolume=1mm3",
            },
            "right",
            {
                "white_surface_total_area_mm^2": 83579.2,
                "mean_thickness_mm": 2.35815,
                "brain_segmentation_volume_mm^3": 1169408.0,
                "brain_segmentation_volume_without_ventricles_mm^3": 1157593.0,
                "brain_segmentation_volume_without_ventricles_from_surf_mm^3": 1157593.0,
                "total_cortical_gray_matter_volume_mm^3": 454587.696158,
                "supratentorial_volume_mm^3": 1023873.0,
                "supratentorial_volume_without_ventricles_mm^3": 1012058.0,
                "estimated_total_intracranial_volume_mm^3": 1420434.160521,
            },
            34,
            [
                {
                    "structure_name": "bankssts",
                    "number_of_vertices": 1094,
                    "surface_area_mm^2": 757,
                    "gray_matter_volume_mm^3": 1725,
                    "average_thickness_mm": 2.215,
                    "thickness_stddev_mm": 0.544,
                    "integrated_rectified_mean_curvature_mm^-1": 0.109,
                    "integrated_rectified_gaussian_curvature_mm^-2": 0.025,
                    "folding_index": 9,
                    "intrinsic_curvature_index": 1.1,
                },
                {
                    "structure_name": "caudalanteriorcingulate",
                    "number_of_vertices": 1137,
                    "surface_area_mm^2": 780,
                    "gray_matter_volume_mm^3": 2327,
                    "average_thickness_mm": 2.842,
                    "thickness_stddev_mm": 0.667,
                    "integrated_rectified_mean_curvature_mm^-1": 0.116,
                    "integrated_rectified_gaussian_curvature_mm^-2": 0.021,
                    "folding_index": 11,
                    "intrinsic_curvature_index": 1.0,
                },
                {
                    "structure_name": "caudalmiddlefrontal",
                    "number_of_vertices": 3126,
                    "surface_area_mm^2": 2218,
                    "gray_matter_volume_mm^3": 5978,
                    "average_thickness_mm": 2.447,
                    "thickness_stddev_mm": 0.605,
                    "integrated_rectified_mean_curvature_mm^-1": 0.122,
                    "integrated_rectified_gaussian_curvature_mm^-2": 0.024,
                    "folding_index": 28,
                    "intrinsic_curvature_index": 3.1,
                },
            ],
        ),
    ],
)
def test_read(
    path,
    headers,
    hemisphere,
    whole_brain_measurements,
    structural_measurements_length,
    structural_measurements_subset,
):
    stats = CorticalParcellationStats.read(path)
    assert headers == stats.headers
    assert hemisphere == stats.hemisphere
    pandas.util.testing.assert_frame_equal(
        left=pandas.DataFrame([whole_brain_measurements]),
        right=stats.whole_brain_measurements,
        check_like=True,  # ignore the order of index & columns
        check_dtype=True,
        check_names=True,
    )
    assert list(stats.structural_measurements.columns) == [
        "structure_name",
        "number_of_vertices",
        "surface_area_mm^2",
        "gray_matter_volume_mm^3",
        "average_thickness_mm",
        "thickness_stddev_mm",
        "integrated_rectified_mean_curvature_mm^-1",
        "integrated_rectified_gaussian_curvature_mm^-2",
        "folding_index",
        "intrinsic_curvature_index",
    ]
    assert len(stats.structural_measurements) == structural_measurements_length
    pandas.util.testing.assert_frame_equal(
        left=pandas.DataFrame(structural_measurements_subset),
        right=stats.structural_measurements.iloc[:3],
        check_like=True,  # ignore the order of index & columns
        check_dtype=True,
        check_names=True,
    )


@pytest.mark.parametrize(
    ("path", "structural_measurements_length"),
    [
        (os.path.join(SUBJECTS_DIR, "soichi", "stats", "lh.BA_exvivo.stats"), 14),
        (
            os.path.join(SUBJECTS_DIR, "soichi", "stats", "lh.BA_exvivo.thresh.stats"),
            14,
        ),
        (os.path.join(SUBJECTS_DIR, "soichi", "stats", "lh.aparc.DKTatlas.stats"), 31),
        (os.path.join(SUBJECTS_DIR, "soichi", "stats", "lh.aparc.a2009s.stats"), 74),
        (os.path.join(SUBJECTS_DIR, "soichi", "stats", "lh.aparc.pial.stats"), 34),
        (os.path.join(SUBJECTS_DIR, "soichi", "stats", "lh.aparc.stats"), 34),
        (os.path.join(SUBJECTS_DIR, "soichi", "stats", "rh.BA_exvivo.stats"), 14),
        (
            os.path.join(SUBJECTS_DIR, "soichi", "stats", "rh.BA_exvivo.thresh.stats"),
            14,
        ),
        (os.path.join(SUBJECTS_DIR, "soichi", "stats", "rh.aparc.DKTatlas.stats"), 31),
        (os.path.join(SUBJECTS_DIR, "soichi", "stats", "rh.aparc.a2009s.stats"), 74),
        (os.path.join(SUBJECTS_DIR, "soichi", "stats", "rh.aparc.pial.stats"), 34),
        (os.path.join(SUBJECTS_DIR, "soichi", "stats", "rh.aparc.stats"), 34),
    ],
)
def test_read_structural_measurements_length(path, structural_measurements_length):
    # simple test to verify no exception gets raised, see test_read for comprehensive test
    stats = CorticalParcellationStats.read(path)
    assert len(stats.structural_measurements) == structural_measurements_length


@pytest.mark.parametrize(
    ("line", "expected_column_name", "expected_value"),
    [
        (
            "Measure Cortex, CortexVol Total cortical gray matter volume, 553998.311189, mm^3",
            "total_cortical_gray_matter_volume_mm^3",
            553998.311189,
        )
    ],
)
def test__parse_whole_brain_measurements_line(
    line, expected_column_name, expected_value
):
    # pylint: disable=protected-access
    column_name, value = CorticalParcellationStats._parse_whole_brain_measurements_line(
        line
    )
    assert column_name == expected_column_name
    assert numpy.allclose(value, [expected_value])


@pytest.mark.parametrize(
    "line",
    ["Measure Cortex, CortexVol Total cortical gray matter volume, 553998.311189"],
)
def test__parse_whole_brain_measurements_line_parse_error(line):
    # pylint: disable=protected-access
    with pytest.raises(ValueError):
        CorticalParcellationStats._parse_whole_brain_measurements_line(line)


@pytest.mark.parametrize(
    "path_str",
    [os.path.join(SUBJECTS_DIR, "fabian", "stats", "lh.aparc.DKTatlas.stats.short")],
)
def test_read_pathlib(path_str: str):
    stats_str = CorticalParcellationStats.read(path_str)
    stats_pathlib = CorticalParcellationStats.read(pathlib.Path(path_str))
    assert stats_str.headers == stats_pathlib.headers


@pytest.mark.parametrize(
    "url",
    [
        "https://raw.githubusercontent.com/fphammerle/freesurfer-stats"
        "/master/tests/subjects/fabian/stats/rh.aparc.stats"
    ],
)
def test_read_https(url: str):
    stats = CorticalParcellationStats.read(url)
    assert stats.headers["generating_program"] == "mris_anatomical_stats"


freesurfer_stats/__init__.py
"""
Python Library to Read FreeSurfer's Cortical Parcellation Anatomical Statistics
([lh]h.aparc(.*)?.stats)

Freesurfer
https://surfer.nmr.mgh.harvard.edu/

>>> from freesurfer_stats import CorticalParcellationStats
>>> stats = CorticalParcellationStats.read('tests/subjects/fabian/stats/lh.aparc.DKTatlas.stats')
>>> stats.headers['CreationTime'].isoformat()
'2019-05-09T21:05:54+00:00'
>>> stats.headers['cvs_version']
'Id: mris_anatomical_stats.c,v 1.79 2016/03/14 15:15:34 greve Exp'
>>> stats.headers['cmdline'][:64]
'mris_anatomical_stats -th3 -mgz -cortex ../label/lh.cortex.label'
>>> stats.hemisphere
>>> stats.whole_brain_measurements['estimated_total_intracranial_volume_mm^3']
0    1.670487e+06
Name: estimated_total_intracranial_volume_mm^3, dtype: float64
>>> stats.whole_brain_measurements['white_surface_total_area_mm^2']
0    98553
Name: white_surface_total_area_mm^2, dtype: int64
>>> stats.structural_measurements[['structure_name', 'surface_area_mm^2',
...                                'gray_matter_volume_mm^3']].head()
            structure_name  surface_area_mm^2  gray_matter_volume_mm^3
0  caudalanteriorcingulate               1472                     4258
1      caudalmiddlefrontal               3039                     8239
2                   cuneus               2597                     6722
3               entorhinal                499                     2379
4                 fusiform               3079                     9064


Copyright (C) 2019 Fabian Peter Hammerle <fabian@hammerle.me>

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
"""

from __future__ import annotations

import datetime
import io
import pathlib
import re
import typing

import numpy
import pandas

from freesurfer_stats.version import __version__


def _get_filepath_or_buffer(
    path: typing.Union[str, pathlib.Path]
) -> typing.Tuple[
    typing.Any, bool  # pandas._typing.FileOrBuffer, bool)
]:  # pragma: no cover
    # can't check coverage due to pandas version branching.
    # pipeline tests against multiple pandas versions.
    if not hasattr(pandas.io.common, "get_filepath_or_buffer"):
        # pandas.io.common.get_filepath_or_buffer was made private in v1.2.0:
        # https://github.com/pandas-dev/pandas/commit/6d1541e1782a7b94797d5432922e64a97934cfa4#diff-934d8564d648e7521db673c6399dcac98e45adfd5230ba47d3aabfcc21979febL247
        # semver?!? breaking change not even mentioned in changelog:
        # https://pandas.pydata.org/pandas-docs/stable/whatsnew/v1.2.0.html
        # new wrapper: get_handle
        # https://github.com/pandas-dev/pandas/blob/v1.2.0/pandas/io/common.py#L490
        # pandas v1.1's get_handle does not yet support urls
        # pylint: disable=no-member; for python<v1.2.0
        io_handle = pandas.io.common.get_handle(path, "r")
        return io_handle.handle, True
    # path_or_buffer: typing.Union[str, pathlib.Path, typing.IO[typing.AnyStr],
    #                              s3fs.S3File, gcsfs.GCSFile]
    # https://github.com/pandas-dev/pandas/blob/v0.25.3/pandas/io/parsers.py#L436
    # https://github.com/pandas-dev/pandas/blob/v0.25.3/pandas/_typing.py#L30
    # pylint: disable=no-member; for python>=v1.2.0
    (path_or_buffer, _, _, should_close) = pandas.io.common.get_filepath_or_buffer(path)
    return path_or_buffer, should_close


class CorticalParcellationStats:

    _HEMISPHERE_PREFIX_TO_SIDE = {"lh": "left", "rh": "right"}
    _GENERAL_MEASUREMENTS_REGEX = re.compile(
        r"^Measure \S+, ([^,\s]+),? ([^,]+), ([\d\.]+), (\S+)$"
    )
    _COLUMN_NAMES_NON_SAFE_REGEX = re.compile(r"\s+")

    def __init__(self):
        self.headers: typing.Dict[str, typing.Union[str, datetime.datetime]] = {}
        self.whole_brain_measurements: typing.Dict[str, typing.Tuple[float, int]] = {}
        self.structural_measurements: typing.Union[pandas.DataFrame, None] = {}

    @property
    def hemisphere(self) -> str:
        return self._HEMISPHERE_PREFIX_TO_SIDE[typing.cast(str, self.headers["hemi"])]

    @staticmethod
    def _read_header_line(stream: typing.TextIO) -> str:
        line = stream.readline()
        assert line.startswith("# ")
        return line[2:].rstrip()

    @classmethod
    def _read_column_header_line(
        cls, stream: typing.TextIO
    ) -> typing.Tuple[int, str, str]:
        line = cls._read_header_line(stream)
        assert line.startswith("TableCol"), line
        line = line[len("TableCol ") :].lstrip()
        index, key, value = line.split(maxsplit=2)
        return int(index), key, value

    def _read_headers(self, stream: typing.TextIO) -> None:
        self.headers = {}
        while True:
            line = self._read_header_line(stream)
            if line.startswith("Measure"):
                break
            if line:
                attr_name, attr_value_str = line.split(" ", maxsplit=1)
                attr_value_str = attr_value_str.lstrip()
                if attr_name in ["cvs_version", "mrisurf.c-cvs_version"]:
                    attr_value = typing.cast(
                        typing.Union[str, datetime.datetime],
                        attr_value_str.strip("$").rstrip(),
                    )
                elif attr_name == "CreationTime":
                    attr_dt = datetime.datetime.strptime(
                        attr_value_str, "%Y/%m/%d-%H:%M:%S-%Z"
                    )
                    if attr_dt.tzinfo is None:
                        assert attr_value_str.endswith("-GMT")
                        attr_dt = attr_dt.replace(tzinfo=datetime.timezone.utc)
                    attr_value = attr_dt
                elif attr_name == "AnnotationFileTimeStamp":
                    attr_value = datetime.datetime.strptime(
                        attr_value_str, "%Y/%m/%d %H:%M:%S"
                    )
                else:
                    attr_value = attr_value_str
                self.headers[attr_name] = attr_value

    @classmethod
    def _format_column_name(cls, name: str, unit: str) -> str:
        column_name = name.lower()
        if unit not in ["unitless", "NA"]:
            column_name += "_" + unit
        return cls._COLUMN_NAMES_NON_SAFE_REGEX.sub("_", column_name)

    @classmethod
    def _parse_whole_brain_measurements_line(
        cls, line: str
    ) -> typing.Tuple[str, numpy.ndarray]:
        match = cls._GENERAL_MEASUREMENTS_REGEX.match(line)
        if not match:
            raise ValueError(f"unexpected line: {line!r}")
        key, name, value, unit = match.groups()
        if (
            key == "SupraTentorialVolNotVent"
            and name.lower() == "supratentorial volume"
        ):
            name += " Without Ventricles"
        column_name = cls._format_column_name(name, unit)
        return column_name, pandas.to_numeric([value], errors="raise")

    @classmethod
    def _read_column_attributes(
        cls, num: int, stream: typing.TextIO
    ) -> typing.List[typing.Dict[str, str]]:
        columns = []
        for column_index in range(1, int(num) + 1):
            column_attrs: typing.Dict[str, str] = {}
            for _ in range(3):
                column_index_line, key, value = cls._read_column_header_line(stream)
                assert column_index_line == column_index
                assert key not in column_attrs
                column_attrs[key] = value
            columns.append(column_attrs)
        return columns

    def _read(self, stream: typing.TextIO) -> None:
        assert (
            stream.readline().rstrip()
            == "# Table of FreeSurfer cortical parcellation anatomical statistics"
        )
        assert stream.readline().rstrip() == "#"
        self._read_headers(stream)
        self.whole_brain_measurements = pandas.DataFrame()
        line = self._read_header_line(stream)
        while not line.startswith("NTableCols"):
            if line.startswith("BrainVolStatsFixed"):
                # https://surfer.nmr.mgh.harvard.edu/fswiki/BrainVolStatsFixed
                assert (
                    line.startswith("BrainVolStatsFixed see ")
                    or line == "BrainVolStatsFixed-NotNeeded because voxelvolume=1mm3"
                )
                self.headers["BrainVolStatsFixed"] = line[len("BrainVolStatsFixed-") :]
            else:
                column_name, value = self._parse_whole_brain_measurements_line(line)
                assert column_name not in self.whole_brain_measurements, column_name
                self.whole_brain_measurements[column_name] = value  # type: ignore
            line = self._read_header_line(stream)
        columns = self._read_column_attributes(int(line[len("NTableCols ") :]), stream)
        assert self._read_header_line(stream) == "ColHeaders " + " ".join(
            c["ColHeader"] for c in columns
        )
        self.structural_measurements = pandas.DataFrame(
            (line.rstrip().split() for line in stream),
            columns=[
                self._format_column_name(c["FieldName"], c["Units"]) for c in columns
            ],
        ).apply(pandas.to_numeric, errors="ignore")

    @classmethod
    def read(cls, path: typing.Union[str, pathlib.Path]) -> CorticalParcellationStats:
        path_or_buffer, should_close = _get_filepath_or_buffer(path)
        stats = cls()
        try:  # pragma: no cover
            # can't check coverage due to pandas version branching.
            # pylint: disable=protected-access; false-positive for ._read
            if isinstance(path_or_buffer, io.TextIOWrapper):  # pandas>=v1.2.0
                stats._read(path_or_buffer)
            elif hasattr(path_or_buffer, "readline"):
                stats._read(io.TextIOWrapper(path_or_buffer))
            else:
                with open(path_or_buffer, "r", encoding="utf8") as stream:
                    stats._read(stream)
        finally:
            if should_close:
                path_or_buffer.close()
        return stats
